chain/consensus/mir/mine.go

```diff
@@ -1,76 +1,159 @@
+//    it is received via reconfiguration channel and applied.
+//
+// There are two ways how membership nodes can be specified or passed to a miner:
+// 1) Environment variable: validators ID and network address are passed via EUDICO_MIR_VALIDATORS variable.
+//    This approach can be used to run Mir in the root network with persistent membership only.
+// 2) Hierarchical consensus framework: ID and network address of joining validators are received via subnet actor state.
+//    The same mechanism is used for reconfiguration. This method is used to run Mir in a subnet.
 func Mine(ctx context.Context, addr address.Address, api v1api.FullNode) error {
 	log.With("addr", addr).Infof("Mir miner started")
 	defer log.With("addr", addr).Infof("Mir miner completed")
 
-	sm, err := NewManager(ctx, addr, api)
+	m, err := NewManager(ctx, addr, api)
 	if err != nil {
-		return fmt.Errorf("unable to create a state manager: %w", err)
+		return fmt.Errorf("unable to create a manager: %w", err)
 	}
-	log := logging.FromContext(ctx, log).With("miner", sm.ID())
+	log := logging.FromContext(ctx, log).With("miner", m.ID())
 
 	log.Infof("Miner info:\n\twallet - %s\n\tnetwork - %s\n\tsubnet - %s\n\tMir ID - %s\n\tvalidators - %v",
-		sm.Addr, sm.NetName, sm.SubnetID, sm.MirID, sm.Validators)
+		m.Addr, m.NetName, m.SubnetID, m.MirID, m.InitialValidatorSet.GetValidators())
+
+	mirErrors := m.Start(ctx)
 
-	mirErrors := sm.Start(ctx)
-	mirHead := sm.App.ChainNotify
+	reconfigure := time.NewTicker(ReconfigurationInterval)
+	defer reconfigure.Stop()
 
-	submit := time.NewTicker(SubmitInterval)
-	defer submit.Stop()
+	lastValidatorSetHash, err := m.InitialValidatorSet.Hash()
+	if err != nil {
+		return err
+	}
 
 	for {
+		// Here we use `ctx.Err()` in the beginning of the `for` loop instead of using it in the `select` statement,
+		// because if `ctx` has been closed then `api.ChainHead(ctx)` returns an error,
+		// and we will be in the infinite loop due to `continue`.
+		if ctx.Err() != nil {
+			log.Debug("Mir miner: context closed")
+			return nil
+		}
 		base, err := api.ChainHead(ctx)
 		if err != nil {
 			log.Errorw("failed to get chain head", "error", err)
 			continue
 		}
 
-		// Miner (leader) for an epoch is assigned deterministically using round-robin.
-		// All other validators use the same Miner in the block.
-		epochMiner := sm.Validators[int(base.Height())%len(sm.Validators)].Addr
 		nextEpoch := base.Height() + 1
 
 		select {
-		case <-ctx.Done():
-			log.With("epoch", nextEpoch).Debug("Mir miner: context closed")
-			return nil
+
 		case err := <-mirErrors:
-			return fmt.Errorf("miner consensus error: %w", err)
-		case batch := <-mirHead:
-			msgs, crossMsgs := sm.GetMessages(batch)
+			// return fmt.Errorf("miner consensus error: %w", err)
+			//
+			// TODO: This is a temporary solution while we are discussing that issue
+			// https://filecoinproject.slack.com/archives/C03C77HN3AS/p1660330971306019
```

Comment:
Did we reach a conclusion?

Reply:
Probably we should use panicking in Mir's Mine function. Otherwise, we need to redesign the subnet.

Reply:
For now, LGTM.